catboost/libs/fstr/shap_values.cpp

```diff
@@ -141,6 +143,117 @@ static void UpdateShapByFeaturePath(
     }
 }
 
+static TContribution SumContributions(
+    const TContribution& lhs,
+    const TContribution& rhs
+) {
+    CB_ENSURE_INTERNAL(
+        lhs.PositiveContribution.size() == rhs.PositiveContribution.size(), 
+        "Contributions have different sizes");
+    TContribution result{lhs.PositiveContribution.size()};
+    const auto approxDimension = result.PositiveContribution.size();
+    for (size_t dimension = 0; dimension < approxDimension; ++dimension) {
+        result.PositiveContribution[dimension] = lhs.PositiveContribution[dimension] + rhs.PositiveContribution[dimension];
+        result.NegativeContribution[dimension] = lhs.NegativeContribution[dimension] + rhs.NegativeContribution[dimension];
+    }
+    return result;
+}
+
+ // Independent Tree SHAP
+TContribution TInternalIndependentTreeShapCalcer::Calc(
+    int depth,
+    size_t nodeIdx,
+    ui32 uniqueFeaturesCount,
+    ui32 featureMatchedDatasetSampleCount
+) {
+    const auto approxDimension = Forest.GetDimensionsCount();
+    const double* leafValuesPtr = Forest.GetFirstLeafPtrForTree(TreeIdx);
```

Comment:
for faster performance, call Forest.GetDimensionsCount(), Forest.GetFirstLeafPtrForTree(TreeIdx), Forest.GetTreeSizes()[TreeIdx] once in calcer's constructor, and store the results in const members, and use these stored results in Calc.

use const array ref to store result of Forest.GetTreeSplits()